### ISSUE TITLE
[Bug] Sapiens2 pointmap is inconsistent across nearly-identical frames — same scene, same pipeline, drastically different surface smoothness

### ISSUE THREAD

**Issue by dafei-qin** [facebookresearch/sapiens2#11]
## Description

When running the 4-task sapiens2 pipeline (`seg → normal → pointmap → pose`) on a video face-crop sequence, the **pointmap output** for the same person / same shot / same pose is unpredictably noisy from frame to frame, while **all other modalities (seg, normal) are stable**.

Two frames that look visually identical to a human can yield pointmaps whose per-pixel neighbour distance differs by ~15×. The resulting meshes range from clean, grid-like quads to heavily crumpled / overlapping quads — the visual defect is so severe it is immediately obvious when rendering the structured mesh (2×2 pixel triangulation).

Most importantly the problem is **not just high-frequency noise**. Per-row analysis shows that in bad frames the predicted X-coordinate of adjacent pixels is **non-monotonic** (neighbouring pixels can swap 3D order), while good frames keep strict left-to-right monotonic X. Temporal / bilateral / Laplacian smoothing all assume "pixel neighbours are 3D neighbours", so they cannot repair this — the output is structurally broken.

Representative overview figure (6 frames × 6 modalities):
- `viz/pointmap_viz_grid.jpg` — columns: X / Y / Z(depth) / pointmap |grad| [0–5 mm] / normal RGB / normal |grad| [0–0.2]. The pointmap gradient column (4th) is **near-black on the two GOOD frames and full of speckle / grid pattern on the 4 bad frames**. Normal and its gradient are almost identical across all 6 frames.
<img width="1152" height="1692" alt="Image" src="https://github.com/user-attachments/assets/6d4d842e-250e-4cce-98bb-f8e812116958" />

Full source frames overview:
- `viz/inputs_compare.jpg` (optional) — orig crop | SR crop | seg mask | depth colormap.
<img width="768" height="1420" alt="Image" src="https://github.com/user-attachments/assets/99ed8023-f088-4de9-bd49-46ab0e88608e" />

Zoom in of predicted point map:
- Good:

<img width="883" height="389" alt="Image" src="https://github.com/user-attachments/assets/c04f9ca8-dd1b-4b30-be01-0c7f9e502753" />

- Bad:
<img width="675" height="472" alt="Image" src="https://github.com/user-attachments/assets/cb9a5ab3-ffd8-4eca-aae1-df0ac6c14173" />

## To Reproduce

Context: sapiens2-1b 4-task pipeline (seg / normal / pointmap / pose). Frames come from a single video face track, RetinaFace-cropped then GFPGAN v1.4 4×-upscaled; all frames run through one A100 in one job.

Steps:

1. `inputs/` contains two 3:4 face crops from the same track `s21_t12`, two seconds apart: `f009060` (good) and `f009120` (bad). They are visually almost indistinguishable to a human.

2. Run sapiens2 pipeline on the two JPGs in the standard way (seg first, then normal/pointmap/pose, pointmap also consumes `seg_dir`). Attached `outputs_good/` and `outputs_bad/` contain what sapiens2 actually produced for these inputs.

3. Minimal reproducer (numpy + plyfile only) to quantify the defect:

```python
import numpy as np
from plyfile import PlyData

def diag(stem, outputs_dir):
    seg = np.load(f'{outputs_dir}/{stem}_seg.npy')
    if seg.ndim == 3: seg = seg.argmax(0)
    mask = seg > 0
    H, W = mask.shape

    # ply stores (n_fg foreground points in seg-mask C-order) + (500 calibration-sphere points)
    ply = PlyData.read(f'{outputs_dir}/{stem}.ply')
    pts = np.stack([ply['vertex']['x'], ply['vertex']['y'], ply['vertex']['z']],
                   axis=1).astype(np.float32)
    pm = np.full((H, W, 3), np.nan, np.float32)
    pm[mask] = pts[:int(mask.sum())]

    # (a) mean 3D distance between image-adjacent pixels (mm)
    dx = pm[:, 1:] - pm[:, :-1]
    d3 = np.linalg.norm(dx, axis=-1)[~np.isnan(dx[..., 0])] * 1000

    # (b) X-coordinate monotonicity along the widest scanline
    row = int(np.argmax(mask.sum(1)))
    Xrow = pm[row, :, 0]
    Xrow = Xrow[~np.isnan(Xrow)]
    dX = np.diff(Xrow) * 1000                   # mm
    print(f'{stem}: dx3d mean={d3.mean():.2f} mm  '
          f'|  row {row}: neg-dX ratio={(dX<0).mean()*100:.1f}%  '
          f'dX std={dX.std():.3f} mm')

diag('s21_t12_f009060', 'outputs_good')    # expect ~0  %, dX std 0.07 mm
diag('s21_t12_f009120', 'outputs_bad')     # expect ~45 %, dX std 1.6  mm
```

Expected numeric output on the attached files:
```
s21_t12_f009060: dx3d mean=0.65 mm  |  neg-dX ratio≈0.0%   dX std≈0.07 mm   <-- GOOD
s21_t12_f009120: dx3d mean=2.70 mm  |  neg-dX ratio≈45.0%  dX std≈1.61 mm   <-- bad
```

4. Applying the per-pixel 2×2 structured-mesh recipe (two tris per 2×2 block, drop faces with max edge > 10 mm) to these two pointmaps produces visibly different meshes — `outputs_good/f009060.obj` is cleanly gridded, `outputs_bad/f009120.obj` has crumpled / overlapping quads.

## Expected Behavior

For near-identical inputs (same subject / pose / shot, only ~1–2 px bbox drift and minor facial micro-motion) the sapiens2 pointmap head should produce temporally **consistent** geometry: adjacent pixels should maintain the 3D ordering implied by their image grid, and per-pixel neighbour 3D distance should stay within a narrow band across the track.

Concretely:
- `pm[i, j+1].x > pm[i, j].x` should hold (strict horizontal monotonicity) when pixels are inside a continuous surface — it does hold on the good frames.
- Neighbour-pixel 3D distance `mean(‖pm[i, j+1] − pm[i, j]‖)` should be stable across the track within ±20%, not vary by 4×.

## Actual Behavior

- `normal` map is **stable** across all 6 frames (per-pixel neighbour angular delta in the same ~0.01 range everywhere; see column 6 of `viz/pointmap_viz_grid.jpg`).
- `seg` is stable (fg ratio, class histogram identical within ±2%).
- Only `pointmap / depth` differ. Absolute `depth` distribution is fine (mean distance to camera 1.30–1.41 m across all frames, face thickness 0.17–0.26 m — plausible in all cases). What breaks is **the local spatial structure of the pointmap**:
  - On good frames (`f009060`, `f009090`), along a horizontal scanline the predicted X coordinates are strictly monotonic (diff stays +0.4 to +3.5 mm, smoothly shrinking toward image centre as expected from perspective projection).
  - On bad frames, the diff sequence is **alternating positive/negative**: `+1.1, −0.1, +1.2, −0.4, +1.4, −0.3, ...`. Adjacent pixels' 3D X ordering is swapped.
  - This is a structural defect in the pointmap head, not additive Gaussian noise. It propagates into any 2×2-block-triangulated mesh as overlapping / self-intersecting quads that look "crumpled".

## Environment

- Sapiens2 commit: 3098c1469fc8ee78c56db9398dedbf30bae0b254
- Python version: 3.13.12 (host analysis / plyfile / pymeshlab); container uses Python 3.12 (torch 2.7.1 + CUDA 12.8) for inference
- PyTorch version: 2.11.0+cu126 (host) / 2.7.1 (inference container)
- CUDA version: 12.6 (host) / 12.8 (inference container); driver 565.57.01 (CUDA 12.7)
- GPU: NVIDIA A100-SXM4-40GB (inference); host analysis is CPU / a different GPU

## Inputs

<img width="772" height="1028" alt="Image" src="https://github.com/user-attachments/assets/ecdc1315-b7e6-4d1c-9b10-6f0131c077ae" />
<img width="768" height="1024" alt="Image" src="https://github.com/user-attachments/assets/4a1478f4-9f5f-4e5e-a630-7b6d6f944cbf" />

**Comment by rawalkhirodkar** [comment]
@dafei-qin sorry you are facing this issue.

- The pointmap model predicts a) per pixel canonical XYZ and b) a scalar, to scale the pointmap to metric.

- Since our models are only trained on image data, we have observed that the predicted scalar can have high variance on a video.

- This can cause the entire pointmap prediction to be jittery (by a scale factor). Very common problem as resolving metric depth from an image is ill posed.

- In this case, I would recommend using light post-processing on the pointmaps across frames to align them into a single coordinate system. Ref: https://github.com/naver/dust3r

Please let me know if this answers your question.

**Comment by dafei-qin** [comment]
Thanks for the prompt of reply!  I'm more thinking of this as a problem on the xyz part instead of the scalar part. Becz if you see from the predicted pointmap, the good one have regular local patterns while the bad one is very noisy.

**Comment by rawalkhirodkar** [comment]
I see. In my experience the xyz being irregular across time steps is rather rare. The model prediction are always faithful to the geometry but can be misaligned on scale across frames.

Unfortunately, we don't have any plans to build a temporal model to resolve this.
